libc/string_catdup.c

```diff
@@ -0,0 +1,780 @@
+{
+	int i;
+	char buf[SIZE_BUF * 2] = NON_ASCII_1;
+	char buf1[] = NON_ASCII_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, strlen(NON_ASCII_1 NON_ASCII_2)));
+	TEST_ASSERT_EQUAL_STRING(NON_ASCII_1 NON_ASCII_2, buf);
+
+	for (i = strlen(NON_ASCII_1 NON_ASCII_2); i < SIZE_BUF * 2; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+/////////////////////////////////////////////////////////////////////////
+
+TEST_SETUP(string_dup)
+{
+}
+
+
+TEST_TEAR_DOWN(string_dup)
+{
+}
+
+
+TEST(string_dup, strdup_basic)
+{
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strdup_empty)
+{
+	char buf[2] = "";
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+	TEST_ASSERT_EQUAL_CHAR('\0', buf1[0]);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strdup_huge_string)
+{
+	int i;
+	char buf[PATH_MAX] = "";
+	char *buf1;
+
+	for (i = 0; i < PATH_MAX - 1; i++) {
+		buf[i] = HUGE_STRING[i];
+	}
+
+	buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf);
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strdup_specials_string)
+{
+	char buf[SIZE_BUF * 2] = SPECIALS_STRING;
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_cat, strdup_ascii)
+{
+	int i;
+	char *buf;
+	char ascii[128] = { 0 };
+
+	for (i = 0; i < 128; i++) {
+		ascii[i] = i;
+	}
+
+	buf = strdup(ascii);
+
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+
+	free(buf);
+}
+
+
+TEST(string_cat, strdup_extended_ascii)
+{
+	int i;
+	char *buf;
+	char ascii[128] = { 0 };
+
+	for (i = 128; i < 255; i++) {
+		ascii[i - 128] = i;
+	}
+
+	buf = strdup(ascii);
+
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+
+	free(buf);
+}
+
+
+TEST(string_dup, strdup_non_ascii)
+{
+	char buf[SIZE_BUF * 2] = NON_ASCII_1 NON_ASCII_2;
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+///////////////////////////////////////////////////////////////////////////////////
+
+TEST(string_dup, strndup_basic)
+{
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strndup(buf, strlen(WORD_1));
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf);
+	TEST_ASSERT_EQUAL_STRING(WORD_1, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strndup_full_string)
+{
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strndup(buf, strlen(buf));
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strndup_zero_size)
+{
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strndup(buf, 0);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf);
+	TEST_ASSERT_EQUAL_STRING("", buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strndup_large_size)
```

Comment:
Could you explain what is the goal of this test case and why the warning is disabled?